### PR TITLE
Option to Add units to the swarm, mesh and associated variables when …

### PR DESCRIPTION
**Pull request by rbeucher**
…saving to HDF5. This replace the Underworld_extended submodule in UWGEO that used to overload some of the main UW classes and was thus prone to bugs...

### CODE REVIEW

File: underworld/mesh/_mesh.py [hunk 1]
```diff
@@ -486,14 +489,20 @@ def _get_iterator(self):
         iset.addAll()
         return iset._get_iterator()
 
-    def save( self, filename ):
+    def save(self, filename, units=None, time=None):
```
Comment (jmansour) [inline underworld/mesh/_mesh.py:492]: @rbeucher can `time` have units, and if so, how does the user set them?

**Comment by jmansour** [comment]
We should update the checkpoint tests to test this new functionality too.

**Comment by julesghub** [comment]
I like the idea of all of this. I agree with @jmansour about updating the tests.

I would also be keen on adding code version information as an attribute to checkpoint files.

File: underworld/mesh/_mesh.py (hunk 1 above)
Comment (rbeucher) [reply underworld/mesh/_mesh.py:492]: Well time is stored as a string. It is here for convenience if you want to save attributes... We could generalize this and allow users to save any attributes...

**Comment by rbeucher** [comment]
What about parsing kwargs and saving the args as attributes...?

File: underworld/mesh/_mesh.py (hunk 1 above)
Comment (jmansour) [reply underworld/mesh/_mesh.py:492]: Okay, let's discuss further at 2pm.

**Comment by rbeucher** [comment]
@julesghub @jmansour give it a go. It should be ready

**Comment by jmansour** [comment]
This is broken for me.  
There's a typo in the `swarm_aux.py` test, and `_swarm.py` is missing an import, but even with those things fix it breaks with

```shell
(default) jmansour@MU00166680:~/code/uw2_rb/docs/test$ python swarm_aux.py
        Global element size: 16x16
        Local offset of rank 0: 0x0
        Local range of rank 0: 16x16
Traceback (most recent call last):
  File "swarm_aux.py", line 93, in <module>
    swarm_save_load('global')
  File "swarm_aux.py", line 62, in swarm_save_load
    clone_swarm.load( "saved_swarm.h5" )
  File "/home/jmansour/code/uw2_rb/underworld/swarm/_swarm.py", line 394, in load
    vals = non_dimensionalise(dset[ chunkStart : chunkEnd ] * units)
numpy.core._exceptions.UFuncTypeError: ufunc 'multiply' did not contain a loop with signature matching types (dtype('<U32'), dtype('<U32')) -> dtype('<U32')
```

File: underworld/mesh/_meshvariable.py [hunk 2]
```diff
@@ -404,10 +413,24 @@ def save( self, filename, meshHandle=None ):
...
+                h5f.attrs['units'] = str(units)
+
+            if time:
+                h5f.attrs["time"] = str(time)
+
             # write to the dset using the global node ids
             local = mesh.nodesLocal
             with dset.collective:
-                dset[mesh.data_nodegId[0:local],:] = self.data[0:local]
+                if units = "degC":
```
Comment (jmansour) [inline underworld/mesh/_meshvariable.py]: @rbeucher why is this special case needed?  It this not directly supported in `pint`?

Comment (rbeucher) [reply underworld/mesh/_meshvariable.py]: no there is an ambiguity around temperature units that needs to be taken care of.

Comment (jmansour) [reply underworld/mesh/_meshvariable.py]: 👍👍

File: underworld/swarm/_swarm.py [hunk 3]
```diff
@@ -334,7 +336,7 @@ def load( self, filename, collective=False, try_optimise=True ):
         with h5File(name=filename, mode="r") as h5f:
 
             try:
-                units = h5f.attrs["units"]
+                units = u.Quantity(h5f.attrs["units"])
```
Comment (jmansour) [inline underworld/swarm/_swarm.py:339]: @rbeucher  why is this handled differently here?

File: docs/test/swarm_aux.py [hunk 4]
```diff
@@ -3,6 +3,14 @@
...
 import underworld as uw
 import numpy as np
+import h5py 
+from underworld.scaling import get_coefficients
+from underworld.scaling import units as u
+from underworld.scaling import non_dimensionalise
+from underworld.scaling import dimensionalise
+
+scaling_coefficients = get_coefficients()
+scaling_coefficients["[length"] = 100 * u.km
```
Comment (jmansour) [inline docs/test/swarm_aux.py:13]: @rbeucher does this do anything?  It doesn't appear to be tested in any case.  I'd suggest pulling it if it isn't tested.

**Comment by jmansour** [comment]
@rbeucher & @julesghub, does it make more sense to add units at the mesh/swarm/variable creation time? This way the objects would carry around their associated units, and checkpointing of the units would be automatic. 

Add reload time there would be potentially a few complications. 

1. If the object was recreated without units, at reload time it'd just adopt the units saved.
2. If recreated with the same units as the checkpoint file, then all is fine and dandy.
3. If recreated with different units to the checkpoint file then we'd trigger an exception. 

Perhaps we can do this for stage 2.

**Comment by jmansour** [comment]
@rbeucher  I've added some minor tweaks, but the `mesh_aux.py` is failing for me. Can you see if it's failing at your end too?  Also, may as well add some quick units tests there too... I'd started doing just that, but then hit that failure, so I'll handball it back to you! 😜

**Comment by rbeucher** [comment]
Ok will have a look! Thanks for the review!

File: docs/test/swarm_aux.py (hunk 4 above)
Comment (rbeucher) [reply docs/test/swarm_aux.py]: No it does not actually, thanks for removing this!

**Comment by rbeucher** [comment]
OK should be fine now but needs the h5oy fix from #522

**Comment by jmansour** [comment]
Feel free to have a look and merge in #522 
There's not much to it.  
Same with #523

**Comment by jmansour** [comment]
Okay, I've made a few more minor tweaks, but tests are passing at this end, and I think it's good to go.

**Pull request merged**